upf/upf_pfcp_server.c

```diff
@@ -1210,9 +1210,24 @@ static uword
 		  (upf_event_urr_hdr_t *) vec_header (uev,
 						      sizeof
 						      (upf_event_urr_hdr_t));
-		upf_session_t *sx;
+		upf_session_t *sx = 0;
+
+		/*
+		 * The session could have been freed or replaced after
+		 * this even has been queued, but before it has been
+		 * handled. So we check if the pool entry is free, and
+		 * also verify that CP-SEID in the session matches
+		 * CP-SEID in the event.
+		 */
+		if (!pool_is_free_index (gtm->sessions, ueh->session_idx))
+		  sx = pool_elt_at_index (gtm->sessions, ueh->session_idx);
+
+		if (!sx || sx->cp_seid != ueh->cp_seid)
+		  {
+		    vec_free_h (uev, sizeof (upf_event_urr_hdr_t));
+		    continue;
+		  }
```

Comment:
Those lines are fine, still I would have coded this as:
```suggestion
		if (!sx || sx->cp_seid != ueh->cp_seid)
		  {
		    goto next_ev_urr;
		  }
```
with a the `goto` label` place right before the free.

That way it would be harder to accidentally use different cleanups for both cases.

Reply:
Updated the code to use `goto`, as indeed we might introduce a hard-to-catch bug later with different cleanup code paths